apps/omg/lib/omg/state/transaction.ex

```diff
@@ -120,3 +116,45 @@ defmodule OMG.State.Transaction do
   def raw_txhash(%__MODULE__.Signed{raw_tx: raw_tx}), do: raw_txhash(raw_tx)
   def raw_txhash(raw_tx), do: hash(raw_tx)
 end
+
+defprotocol OMG.State.Transaction.Protocol do
+  @moduledoc """
+  Should be implemented for any type of transaction processed in the system
+  """
+
+  alias OMG.InputPointer
+  alias OMG.Output
+  alias OMG.State.Transaction
+
+  @doc """
+  Transforms structured data into RLP-structured (encodable) list of fields
+  """
+  @spec get_data_for_rlp(t()) :: list(any())
+  def get_data_for_rlp(tx)
+
+  @doc """
+  List of input pointers (e.g. of which one implementation is `utxo_pos`) this transaction is intending to spend
+  """
+  @spec get_inputs(t()) :: list(InputPointer.Protocol.t())
+  def get_inputs(tx)
+
+  @doc """
+  List of outputs this transaction intends to create
+  """
+  @spec get_outputs(t()) :: list(Output.Protocol.t())
+  def get_outputs(tx)
+
+  @doc """
+  Custom validation of the transaction with respect to its witnesses. Part of stateless validation routine
+  """
+  @spec valid?(t(), Transaction.Signed.t()) :: true | {:error, atom}
+  def valid?(tx, signed_tx)
+
+  @doc """
+  Custom stateful validity, based on pre-fetched subset of input UTXOs
```

Comment:
The meaning is 'checks if transaction can be applied to given input utxos`? If I'm right I would put that in the doc.